File: python/agent/loop.py
"""101: the agentic loop.

    python -m agent.loop "What time is it? Remember it under last_run."

One call to the model, run whatever tools it asked for, feed results back, repeat
until it stops asking. Everything that happened lands in runs/<id>.json.

The SDK ships a helper that does this loop for you (client.beta.messages.tool_runner).
It is written out here because the loop is the point of 101.
"""
import json
import os
import sys
import time
import uuid
from pathlib import Path

import anthropic

from agent import RUNS_DIR, tools

MODEL = os.environ.get("MODEL", "claude-sonnet-5")
SYSTEM = (Path(__file__).parent / "system.md").read_text()
MAX_TURNS = int(os.environ.get("MAX_TURNS", "10"))

client = anthropic.Anthropic()


def run(task: str, on_event=None) -> dict:
    def emit(event: dict) -> None:
        if on_event:
            on_event(event)

    run_id = time.strftime("%Y%m%d-%H%M%S-") + uuid.uuid4().hex[:6]
    messages = [{"role": "user", "content": task}]
    trace = {"id": run_id, "task": task, "model": MODEL, "steps": [], "usage": {"input": 0, "output": 0}}

    emit({"type": "start", "run_id": run_id, "task": task})

    # ─── THE LOOP ────────────────────────────────────────────────────────────
    for turn in range(MAX_TURNS):
        emit({"type": "thinking", "turn": turn + 1})   # ← loop back to model

        response = client.messages.create(
            model=MODEL,
            max_tokens=16000,
            system=SYSTEM,
            tools=tools.SCHEMAS,
            messages=messages,
        )
        trace["usage"]["input"] += response.usage.input_tokens
        trace["usage"]["output"] += response.usage.output_tokens
        messages.append({"role": "assistant", "content": response.content})

        if response.stop_reason != "tool_use":
            break                                       # ← model decided it's done

        results = []
        for block in response.content:
            if block.type != "tool_use":
                continue
            emit({"type": "tool_call", "turn": turn + 1, "tool": block.name, "input": block.input})
            out = tools.dispatch(block.name, block.input)
            truncated = out[:2000]
            trace["steps"].append({"tool": block.name, "input": block.input, "output": truncated})
            emit({"type": "tool_result", "turn": turn + 1, "tool": block.name, "output": truncated})
            results.append(
                {
                    "type": "tool_result",
                    "tool_use_id": block.id,
                    "content": out,
                    "is_error": out.startswith("error:"),
                }
            )
        messages.append({"role": "user", "content": results})  # ← feed results back, repeat
    # ─────────────────────────────────────────────────────────────────────────

    # stop_reason still "tool_use" here means MAX_TURNS was hit
    trace["stop_reason"] = response.stop_reason
    trace["output"] = "".join(b.text for b in response.content if b.type == "text")
    save(trace)
    emit({"type": "answer", "text": trace["output"], "run_id": run_id,
          "usage": trace["usage"], "stop_reason": trace["stop_reason"],
          "step_count": len(trace["steps"])})
    return trace


def save(trace: dict) -> Path:
    RUNS_DIR.mkdir(parents=True, exist_ok=True)
    path = RUNS_DIR / f"{trace['id']}.json"
    path.write_text(json.dumps(trace, indent=2, default=str))
    return path


if __name__ == "__main__":
    task = " ".join(sys.argv[1:]) or "What time is it? Remember it under last_run."
    t = run(task)
    print(t["output"])
    print(
        f"\n[{len(t['steps'])} tool calls, stop={t['stop_reason']}, "
        f"tokens in/out={t['usage']['input']}/{t['usage']['output']}, trace={RUNS_DIR}/{t['id']}.json]",
        file=sys.stderr,
    )

File: ts/src/loop.ts
// 101: the agentic loop.
//
//   npm run loop -- "What time is it? Remember it under last_run."
//
// One call to the model, run whatever tools it asked for, feed results back, repeat
// until it stops asking. Everything that happened lands in runs/<id>.json.
//
// The SDK ships a helper that does this loop for you (client.beta.messages.toolRunner).
// It is written out here because the loop is the point of 101.
import fs from "node:fs";
import path from "node:path";
import Anthropic from "@anthropic-ai/sdk";
import { MODEL, RUNS_DIR, here } from "./env";
import * as tools from "./tools";

const SYSTEM = fs.readFileSync(path.join(here, "system.md"), "utf8");
const MAX_TURNS = 10;

export let client = new Anthropic();
export function setClient(c: Anthropic) {
  client = c;
}

export type Trace = {
  id: string;
  task: string;
  model: string;
  steps: { tool: string; input: unknown; output: string }[];
  usage: { input: number; output: number };
  stop_reason?: string | null;
  output?: string;
  verdict?: Record<string, unknown>;
};

export async function run(task: string): Promise<Trace> {
  const id = stamp() + "-" + Math.random().toString(16).slice(2, 8);
  const messages: Anthropic.MessageParam[] = [{ role: "user", content: task }];
  const trace: Trace = { id, task, model: MODEL, steps: [], usage: { input: 0, output: 0 } };

  let response!: Anthropic.Message;
  for (let turn = 0; turn < MAX_TURNS; turn++) {
    response = await client.messages.create({
      model: MODEL,
      max_tokens: 16000,
      system: SYSTEM,
      tools: tools.SCHEMAS,
      messages,
    });
    trace.usage.input += response.usage.input_tokens;
    trace.usage.output += response.usage.output_tokens;
    messages.push({ role: "assistant", content: response.content });

    if (response.stop_reason !== "tool_use") break;

    const results: Anthropic.ToolResultBlockParam[] = [];
    for (const block of response.content) {
      if (block.type !== "tool_use") continue;
      const out = await tools.dispatch(block.name, block.input);
      trace.steps.push({ tool: block.name, input: block.input, output: out.slice(0, 2000) });
      results.push({ type: "tool_result", tool_use_id: block.id, content: out, is_error: out.startsWith("error:") });
    }
    messages.push({ role: "user", content: results });
  }

  // stop_reason still "tool_use" here means MAX_TURNS was hit
  trace.stop_reason = response.stop_reason;
  trace.output = response.content.filter((b) => b.type === "text").map((b) => b.text).join("");
  save(trace);
  return trace;
}

export function stamp(): string {
  const d = new Date().toISOString().replace(/[-:T]/g, "").slice(0, 14);
  return `${d.slice(0, 8)}-${d.slice(8)}`;
}

export function save(trace: Trace): string {
  fs.mkdirSync(RUNS_DIR, { recursive: true });
  const file = path.join(RUNS_DIR, `${trace.id}.json`);
  fs.writeFileSync(file, JSON.stringify(trace, null, 2));
  return file;
}

if (process.argv[1]?.endsWith("loop.ts")) {
  const task = process.argv.slice(2).join(" ") || "What time is it? Remember it under last_run.";
  const t = await run(task);
  console.log(t.output);
  console.error(
    `\n[${t.steps.length} tool calls, stop=${t.stop_reason}, tokens in/out=${t.usage.input}/${t.usage.output}, trace=${RUNS_DIR}/${t.id}.json]`,
  );
}
